ir/program/program.rs

```diff
@@ -4,20 +4,28 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use std::collections::HashMap;
+use std::sync::Arc;
 
 use encoding::graph::definition::definition_key::DefinitionKey;
 
-use crate::program::{block::FunctionalBlock, function::FunctionIR};
+use crate::{
+    inference::type_inference::{FunctionAnnotations, TypeAnnotations},
+    program::{
+        block::FunctionalBlock,
+        function::{FunctionDefinitionError, FunctionIR},
+        function_headers::{FunctionID, FunctionIDTrait, FunctionIndex},
+    },
+};
 
 pub struct Program {
     entry: FunctionalBlock,
-    functions: HashMap<DefinitionKey<'static>, FunctionIR>,
+    functions: Vec<FunctionIR>,
 }
 
 impl Program {
-    pub fn new(entry: FunctionalBlock, functions: HashMap<DefinitionKey<'static>, FunctionIR>) -> Self {
+    pub fn new(entry: FunctionalBlock, functions: Vec<FunctionIR>) -> Self {
         // TODO: verify exactly the required functions are provided
+        // TODO: ^ Why? I've since interpreted it as the query-local functions
```

Comment:
I interpret the functions passed to a program as the IR of the preamble functions only.
An `AnnotatedProgram` has access to all the functions in the schema as well.